### ISSUE TITLE
Error with multiple requirements files, one containing an --extra-index-url

### ISSUE THREAD

**Issue by davidhao3300** [astral-sh/uv#2840]
Consider the following contents:
req1.txt: `filelock==3.12.4`
req2.txt: 
```
--extra-index-url https://download.pytorch.org/whl/cpu

torch==2.0.1
```

Running `uv pip install -r req1.txt -r req2.txt` fails with
```
  × No solution found when resolving dependencies:
  ╰─▶ Because there is no version of filelock==3.12.4 and you require filelock==3.12.4, we can conclude that the requirements are
      unsatisfiable.
```

It appears to be the --extra-index-url that causes issues, removing fixes the issue.

**Comment by charliermarsh** [comment]
Thanks, explained in detail here including ways to fix it: https://github.com/astral-sh/uv/blob/main/PIP_COMPATIBILITY.md#packages-that-exist-on-multiple-indexes.

**Comment by davidhao3300** [comment]
ah, thank you very much, thanks for the info. From reading this, it appears that we expect uv to first check the extra index, before falling back to the default index?

Does this mean the torch index contains filelock in it, but not the version we need?

**Comment by davidhao3300** [comment]
Appears to be the case - https://download.pytorch.org/whl/filelock/

**Comment by charliermarsh** [comment]
That's correct -- the Torch index only contains 3.9.0: https://download.pytorch.org/whl/filelock/

**Comment by charliermarsh** [comment]
You should be able to pass `--index-strategy unsafe-any-match` if you want to allow `uv` to look at versions across any index. Or you can use 3.9.0 :)

**Comment by davidhao3300** [comment]
Okay! I think the workaround is sufficient, and swap over to https://github.com/astral-sh/uv/issues/171 when it's available, thank you for the incredibly fast response time!
